chain/chunks/src/lib.rs

```diff
@@ -909,6 +906,39 @@ impl ShardsManager {
         Ok(header)
     }
 
+    pub fn insert_chunk_forward(&mut self, forward: PartialEncodedChunkForwardMsg) {
+        let chunk_hash = forward.chunk_hash.clone();
+        let num_total_parts = self.runtime_adapter.num_total_parts() as u64;
+        match self.chunk_forwards_cache.cache_get_mut(&chunk_hash) {
+            None => {
+                // Never seen this chunk hash before, collect the parts and cache them
+                let parts = forward.parts.into_iter().filter_map(|part| {
+                    let part_ord = part.part_ord;
+                    if part_ord > num_total_parts {
+                        warn!(target: "chunks", "Received chunk part with part_ord greater than the the total number of chunks");
+                        None
+                    } else {
+                        Some((part_ord, part))
+                    }
+                }).collect();
+                self.chunk_forwards_cache.cache_set(chunk_hash, parts);
+            }
+
+            Some(existing_parts) => {
+                for part in forward.parts {
+                    let part_ord = part.part_ord;
+                    if part_ord > num_total_parts {
+                        warn!(target: "chunks", "Received chunk part with part_ord greater than the the total number of chunks");
+                        continue;
+                    }
+                    if let Some(_) = existing_parts.insert(part_ord, part) {
```

Comment:
This doesn't work naively, even if we do check signatures of the block producers (see my comment re: signatures below).

A malicious actor with a network connection better than ours can quickly observe that a new chunk is produced in the network, and send us the chunk hash with a bunch of invalid chunk parts. We will cache their chunk parts, and when the actual forwarded messages with valid parts arrive, we will not update them here.

By the time we collected all the parts, and we call to `process_partial_encoded_chunk`, the parts will be rejected due to proofs not matching, but we will incur the 1s delay. The adversary can execute this attack at every height, significantly slowing down the system.

To address it, we need to send the merkle root (merkle proofs are already included in the parts here), but we also need to prove that the merkle root corresponds to the chunk hash. To avoid including the entire chunk header, we can change the way we compute the chunk hash to:

```
chunk_header_inner = hash(borsh_serialize(header)); // this is the hash today
chunk_header_new = hash(merkle_root || chunk_header_inner);
```

With this new logic (we can also optionally exclude merkle root from the computation of `inner`, but I don't think the extra work to do that is worth it) the `forward` message will contain:

```
chunk_hash_inner: CryptoHash,
merkle_root: CryptoHash,
signature: Signature
<some other info to verify the sig, potentially just height_created>
```

Using this info the node can
a) Compute the actual chunk hash by concatting and hashing the two hashes given
b) Verify the signature against the computed hash.
c) Verify the merkle proofs against the merkle root.

(see my comment below re: necessity to include the signature)

Reply:
> when the actual forwarded messages with valid parts arrive, we will not update them here

They are updated here. The `insert` call still happens; what is returned is the value that used to be under the key before the `insert` call (if any). The attacker would need to time the messages right so that their faulty message came in after the real one. This might still be doable though, so it is probably still good to include the verification.

I did think about the possibility of invalid forwarded chunks, but since the failure mode is falling back on the existing request/response chunk distribution, it didn't seem that bad. Though you are probably right that an attack which is essentially free to execute cannot be left open, no matter how minor its impact.

Reply:
Validation added in https://github.com/nearprotocol/nearcore/pull/2769/commits/54c6a86343154c76182a5ebba57c09e496d44d3a